### PR TITLE
web: analytics for resource pinning

### PR DESCRIPTION
**Pull request by landism**

### CODE REVIEW

**Review by nicks — APPROVED** [review]

File: web/src/SidebarPin.tsx [hunk 1]
```diff
@@ -49,18 +50,24 @@ export function SidebarPinContextProvider(
...
 
   function unpinResource(name: string) {
-    setPinnedResources(prevState =>
-      !prevState ? prevState : prevState.filter(n => n !== name)
-    )
+    setPinnedResources(prevState => {
+      const ret = !prevState ? prevState : prevState.filter(n => n !== name)
+      incr("ui.web.pin", {
+        newPinCount: ret.length.toString(),
+        pinning: "false",
```
Comment (maiamcc) [inline web/src/SidebarPin.tsx:67]: i'm not a huge fan of this but idk what else to change it to. i assume `true` -> it was a "pin" action and `false` -> it was an "unpin" action? this sorta reads to me as "anything is pinned: t/f". but also idk if there's a better name so eh.

**Review by maiamcc — APPROVED** [review]

File: web/src/SidebarPin.tsx [hunk 2]
```diff
@@ -49,18 +50,24 @@ export function SidebarPinContextProvider(
...
   }
 
   function unpinResource(name: string) {
-    setPinnedResources(prevState =>
-      !prevState ? prevState : prevState.filter(n => n !== name)
-    )
+    setPinnedResources(prevState => {
+      const ret = !prevState ? prevState : prevState.filter(n => n !== name)
+      incr("ui.web.pin", {
+        newPinCount: ret.length.toString(),
```
[the remark this answers is not in the document]
Comment (landism) [reply web/src/SidebarPin.tsx:66]: Good catch!

That is a reasonable inference from line 64. My gut reaction was "well if that was a problem, the type checker would have caught it!" It turns out that we have `strict` enabled in our ts config, so `prevState`, which is a `string[]`, cannot be equal to `null` or `undefined`, so line 64 should really just be `const ret = prevState.filter(n => n !== name)` and the use of `NonNullable` above is totally gratuitous.

File: web/src/SidebarPin.tsx (hunk 1 above)
Comment (landism) [reply web/src/SidebarPin.tsx:67]: yeahhhh. I originally didn't have this tag and named the actions "ui.web.pin" and "ui.web.unpin", but I figured it'd be better for them to share a name to make querying for # of pinned items easier. Then I was going to go with `pinning: "true"` and `unpinning: "true"`, but that seems silly. `action: "pin"` and `action: "unpin"` seem a lot better, so I'll go with that. Thanks for prodding on this!

File: web/src/SidebarResources.test.tsx [hunk 3]
```diff
@@ -33,6 +35,15 @@ function clickPin(
...
+  beforeEach(() => {
+    fetchMock.resetMocks()
+    fetchMock.mockResponse(JSON.stringify({}))
+  })
+
+  afterEach(() => {
+    fetchMock.resetMocks()
+  })
+
   it("adds items to the pinned group when items are pinned", () => {
```
[the remark this answers is not in the document]
Comment (landism) [reply web/src/SidebarResources.test.tsx:47]: why mess with perfection?

[the remark this answers is not in the document]
Comment (maiamcc) [reply web/src/SidebarResources.test.tsx:47]: 🤔 🤔 🤔 🤔 🤔

**Pull request merged**